orchagent/pfcwdorch.cpp

```diff
@@ -55,28 +58,36 @@ void PfcWdOrch<DropHandler, ForwardHandler>::doTask(Consumer& consumer)
         return;
     }
 
-    auto it = consumer.m_toSync.begin();
-    while (it != consumer.m_toSync.end())
+    if ((consumer.getDbId() == CONFIG_DB) && (consumer.getTableName() == CFG_PFC_WD_TABLE_NAME))
     {
-        KeyOpFieldsValuesTuple t = it->second;
+        auto it = consumer.m_toSync.begin();
+        while (it != consumer.m_toSync.end())
+        {
+            KeyOpFieldsValuesTuple t = it->second;
 
-        string key = kfvKey(t);
-        string op = kfvOp(t);
+            string key = kfvKey(t);
+            string op = kfvOp(t);
 
-        if (op == SET_COMMAND)
-        {
-            createEntry(key, kfvFieldsValues(t));
-        }
-        else if (op == DEL_COMMAND)
-        {
-            deleteEntry(key);
+            if (op == SET_COMMAND)
+            {
+                createEntry(key, kfvFieldsValues(t));
+            }
+            else if (op == DEL_COMMAND)
+            {
+                deleteEntry(key);
+            }
+            else
+            {
+                SWSS_LOG_ERROR("Unknown operation type %s\n", op.c_str());
+            }
+
+            consumer.m_toSync.erase(it++);
         }
-        else
+
+        if (consumer.m_toSync.empty())
         {
-            SWSS_LOG_ERROR("Unknown operation type %s\n", op.c_str());
+            m_entriesCreated = true;
```

Comment:
>m_entriesCreated [](start = 12, length = 16)

why you need this?

Reply:
This is introduced because we have two separate consumers now (as per your requested change), one in APPL_DB with table name PFC_WD_TABLE as the key in m_consumerMap, and one in COUNTERS_DB with table name PFC_WD as the key.

This introduces dependency between the execution of the two tables. PFC_WD should come first to startWdOnPort() per queue, before we can startWdActionOnQueue, where PFC_WD_TABLE keeps the in-storm queues on the port, where drop actions need to be taken further.

However, because the consumers are maintained as a key-value map, there is no guarantee which consumer comes in first. Introducing m_entriesCreated serves as a barrier in this case to ensure that PFC_WD consumer is processed before that of PFC_WD_TABLE so that we do not waste CPU cycles iterating all in-storm queues  in PFC_WD_TABLE without being able to start the drop action and retry a second time, although executing PFC_WD_TABLE before PFC_WD is totally fine from the code logic perspective (startWdActionOnQueue will return false and continue).


The second case this variable can help is when ports to start watchdog on comes in one batch in PFC_WD. This is the only case as far as I observe during both the cold- and warm-reboot initializations. m_entriesCreated ensures that PFC_WD is processed to its completion before PFC_WD_TABLE is ever processed. This guarantees that when PFC_WD_TABLE is handled, all lossless queues of all ports have watchdog started, and we will succeed starting action on any lossless queue without a doTask retry execution.

It should be totally fine removing the variable, and rely solely on doTask retries to resolve the dependency to converge all the in-storm queues into the drop action state untimately. But for the sake of efficiency, introducing the variable ensures some level of execution order and gives us better chances to successfully start action on all queues in one round of doTask.